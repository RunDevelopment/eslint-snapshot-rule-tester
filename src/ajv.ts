/**
 * @fileoverview The instance of Ajv validator.
 * @author Evgeny Poberezkin
 * @license MIT
 */

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

import Ajv from "ajv"
import metaSchema from "ajv/lib/refs/json-schema-draft-04.json"

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

export default (additionalOptions: Partial<Ajv.Options> = {}): Ajv.Ajv => {
    const ajv = new Ajv({
        meta: false,
        useDefaults: true,
        validateSchema: false,
        missingRefs: "ignore",
        verbose: true,
        schemaId: "auto",
        ...additionalOptions,
    })

    ajv.addMetaSchema(metaSchema)
    // eslint-disable-next-line no-underscore-dangle -- Ajv's API
    ajv._opts["defaultMeta" as never] = metaSchema.id as never

    return ajv
}
