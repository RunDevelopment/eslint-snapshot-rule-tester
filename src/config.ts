import type { Linter } from "eslint"
import { builtinRules } from "eslint/use-at-your-own-risk"

export const defaultConfig = [
    {
        plugins: {
            "@": {
                /*
                 * Because we try to delay loading rules until absolutely
                 * necessary, a proxy allows us to hook into the lazy-loading
                 * aspect of the rules map while still keeping all of the
                 * relevant configuration inside of the config array.
                 */
                rules: new Proxy(
                    {},
                    {
                        get(_target, property) {
                            if (typeof property === "string") {
                                return builtinRules.get(property)
                            }
                            return undefined
                        },
                        has(_target, property) {
                            return typeof property === "string" && builtinRules.has(property)
                        },
                    },
                ),
            },
        },
        languageOptions: {
            sourceType: "module",
            ecmaVersion: "latest",
            parser: require("espree"),
            parserOptions: {},
        },
        linterOptions: {
            reportUnusedDisableDirectives: 1,
        },
    },

    // default ignores are listed here
    {
        ignores: ["**/node_modules/", ".git/"],
    },

    // intentionally empty config to ensure these files are globbed by default
    {
        files: ["**/*.js", "**/*.mjs"],
    },
    {
        files: ["**/*.cjs"],
        languageOptions: {
            sourceType: "commonjs",
            ecmaVersion: "latest",
        },
    },
] satisfies Linter.FlatConfig[]
