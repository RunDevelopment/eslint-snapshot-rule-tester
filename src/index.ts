export { TestCase, TestCases } from "./test-case"
export { SnapshotRuleTester } from "./rule-tester"
