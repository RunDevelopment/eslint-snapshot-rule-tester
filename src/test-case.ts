import type { Linter, RuleTester } from "eslint"

export function toTestCase(testCase: TestCase | string): TestCase {
    if (typeof testCase === "string") {
        return { code: testCase }
    }
    return testCase
}

export interface TestCase extends RuleTester.ValidTestCase, Partial<Linter.FlatConfig> {
    options?: unknown[]
    settings?: Record<string, unknown>
}

export interface TestCases {
    valid: (TestCase | string)[]
    invalid: (TestCase | string)[]
}

/*
 * List every parameters possible on a test case that are not related to eslint
 * configuration
 */
export const ruleTesterParameters = [
    "name",
    "code",
    "filename",
    "options",
    "only",
] satisfies (keyof TestCase)[]
