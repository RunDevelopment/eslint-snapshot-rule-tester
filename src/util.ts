import type { Rule } from "eslint"
import type { JSONSchema4 } from "json-schema"

/**
 * Replace control characters by `\u00xx` form.
 * @param text The text to sanitize.
 * @returns The sanitized text.
 */
export function sanitize(text: string): string {
    if (typeof text !== "string") {
        return ""
    }
    return text.replace(
        /[\u0000-\u0009\u000b-\u001a]/gu, // eslint-disable-line no-control-regex -- Escaping controls
        (c) => `\\u${c.codePointAt(0)!.toString(16).padStart(4, "0")}`,
    )
}
/**
 * Clones a given value deeply.
 * Note: This ignores `parent` property.
 * @param x A value to clone.
 * @returns A cloned value.
 */
export function cloneDeeplyExcludesParent(x: unknown): unknown {
    if (typeof x === "object" && x !== null) {
        if (Array.isArray(x)) {
            return x.map(cloneDeeplyExcludesParent)
        }

        const retv: Record<string, unknown> = {}

        for (const key in x) {
            if (key !== "parent" && Object.hasOwn(x, key)) {
                retv[key] = cloneDeeplyExcludesParent(x[key as never])
            }
        }

        return retv
    }

    return x
}

/**
 * Freezes a given value deeply.
 * @param x A value to freeze.
 * @returns
 */
export function freezeDeeply(x: unknown): void {
    if (typeof x === "object" && x !== null) {
        if (Array.isArray(x)) {
            x.forEach(freezeDeeply)
        } else {
            for (const key in x) {
                if (key !== "parent" && Object.hasOwn(x, key)) {
                    freezeDeeply(x[key as never])
                }
            }
        }
        Object.freeze(x)
    }
}

/**
 * Gets a complete options schema for a rule.
 * @param rule A new-style rule object
 * @returns JSON Schema for the rule's options.
 */
export function getRuleOptionsSchema(rule: Rule.RuleModule): JSONSchema4 | null {
    const schema = rule.schema || (rule.meta && rule.meta.schema)

    if (Array.isArray(schema)) {
        if (schema.length) {
            return {
                type: "array",
                items: schema,
                minItems: 0,
                maxItems: schema.length,
            }
        }
        return {
            type: "array",
            minItems: 0,
            maxItems: 0,
        }
    }

    // Given a full schema, leave it alone
    return schema || null
}

export function splitLines(s: string): string[] {
    return s.split(/\r?\n/)
}

export function capitalize(s: string): string {
    return s.charAt(0).toUpperCase() + s.slice(1)
}

export function isObject(obj: unknown): obj is NonNullable<object> {
    return typeof obj === "object" && obj !== null
}

export function includesInvalidSurrogatePair(s: string): boolean {
    const withoutValid = s.replace(/[\uD800-\uDBFF][\uDC00-\uDFFF]/g, "")
    return /[\uD800-\uDFFF]/.test(withoutValid)
}
