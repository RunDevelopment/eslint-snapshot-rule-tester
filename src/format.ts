import type { Linter } from "eslint"
import { TestCase } from "./test-case"
import YAML, { ToStringOptions } from "yaml"
import { capitalize, splitLines } from "./util"

const yamlOptions: ToStringOptions = {
    blockQuote: "literal",
}

export function format(
    testCase: TestCase,
    fixable: boolean,
    output: string | undefined,
    messages: readonly Linter.LintMessage[],
): string {
    let result = ""

    result += formatConfig(testCase)
    result += formatCode(testCase, messages)
    result += formatOutput(output, fixable)
    result += formatMessages(testCase, messages)

    return result.replace(/\r\n/g, "\n").replace(/\n{0,3}$/, "")
}
const ignoreKeys = new Set<keyof TestCase>(["code", "name"])
function formatConfig(testCase: TestCase): string {
    const copyKeys = (Object.keys(testCase) as (keyof TestCase)[])
        .filter((key) => !ignoreKeys.has(key))
        .sort()
    const configObject: Record<string, unknown> = {}
    let hasConfig = false
    for (const key of copyKeys) {
        const value = testCase[key]
        if (value !== undefined) {
            configObject[capitalize(key)] = value
            hasConfig = true
        }
    }

    if (!hasConfig) {
        return ""
    }
    try {
        return `${YAML.stringify(configObject, yamlOptions)}\n`
    } catch {
        // this can happen, so we just ignore the keys causing the error
        for (const key of Object.keys(configObject)) {
            const value = configObject[key]
            try {
                YAML.stringify(value, yamlOptions)
            } catch {
                configObject[key] = "unable to serialize"
            }
        }

        return `${YAML.stringify(configObject, yamlOptions)}\n`
    }
}

interface DecodedLoc {
    start: number
    startContinued: boolean
    end: number
    endContinued: boolean
    messageIndex: number
}
function groupByLine(
    codeLines: readonly string[],
    messages: readonly Linter.LintMessage[],
): Map<number, DecodedLoc[]> {
    interface MessageLocation {
        line: number
        column: number
        endLine: number
        endColumn: number
        messageIndex: number
    }

    const locations = messages
        .map((message, i): MessageLocation => {
            return {
                line: message.line,
                column: message.column,
                endLine: message.endLine ?? message.line,
                endColumn: message.endColumn ?? message.column + 1,
                messageIndex: i,
            }
        })
        .sort((a, b) => {
            return (
                a.line - b.line ||
                a.column - b.column ||
                b.endLine - a.endLine ||
                b.endColumn - a.endColumn
            )
        })

    function decodeLoc(loc: MessageLocation, line: number, length: number): DecodedLoc {
        let start, startContinued, end, endContinued
        if (loc.line < line) {
            start = 0
            startContinued = true
        } else {
            // loc.line === line
            start = loc.column - 1
            startContinued = false
        }
        if (loc.endLine > line) {
            end = length
            endContinued = true
        } else {
            // loc.endLine === line
            end = loc.endColumn - 1
            endContinued = false
        }
        return { start, startContinued, end, endContinued, messageIndex: loc.messageIndex }
    }

    const byLines = new Map<number, DecodedLoc[]>()
    for (const loc of locations) {
        for (let line = loc.line; line <= loc.endLine; line++) {
            const list = byLines.get(line) ?? []
            const lineLength = codeLines.at(line - 1)?.length ?? 0
            list.push(decodeLoc(loc, line, lineLength))
            byLines.set(line, list)
        }
    }

    return byLines
}
function standardFormatLocations(locations: readonly DecodedLoc[], offset: number): string[] {
    const locLines: string[] = []

    for (const loc of locations) {
        const { start, startContinued, end, endContinued, messageIndex } = loc

        const content =
            (startContinued ? "~" : "^") +
            "~".repeat(Math.max(0, end - start - 1)) +
            (endContinued ? `\\ [${messageIndex + 1}] ` : ` [${messageIndex + 1}] `)

        const locLineIndex = offset + start

        // try to fit it into any previous lines
        let added = false
        for (let i = 0; i < locLines.length; i++) {
            const line = locLines[i]
            if (line.length <= locLineIndex) {
                locLines[i] = line.padEnd(locLineIndex) + content
                added = true
                break
            }
        }

        if (!added) {
            locLines.push("".padEnd(locLineIndex) + content)
        }
    }

    return locLines
}
function compactFormatLocations(
    locations: readonly DecodedLoc[],
    offset: number,
): string[] | undefined {
    let squiggleLine = ""
    let messageLine = ""

    for (const [loc, index] of locations.map((loc, i) => [loc, i] as const)) {
        const nextLoc = locations.at(index + 1)

        const { start, startContinued, end, endContinued, messageIndex } = loc
        if (startContinued || endContinued) {
            return undefined
        }

        const squiggle = "^" + "~".repeat(Math.max(0, end - start - 1))
        const num = `[${messageIndex + 1}]`
        // unindent if it means that we can fit the next location
        const unindentNum = nextLoc && nextLoc.start === start + num.length

        if (squiggleLine.length <= start) {
            squiggleLine = squiggleLine.padEnd(start) + squiggle + " "
        } else {
            return undefined
        }
        if (unindentNum && messageLine.length <= start - 1) {
            messageLine = messageLine.padEnd(start - 1) + num + " "
        } else if (messageLine.length <= start) {
            messageLine = messageLine.padEnd(start) + num + " "
        } else {
            return undefined
        }
    }

    return ["".padEnd(offset) + squiggleLine, "".padEnd(offset) + messageLine]
}
function formatCode(testCase: TestCase, messages: readonly Linter.LintMessage[]): string {
    const codeLines = splitLines(testCase.code)
    const byLines = groupByLine(codeLines, messages)
    const outputLines: string[] = []

    codeLines.forEach((codeLine, lineIndex) => {
        const prefix = "  "
        const offset = prefix.length
        outputLines.push(prefix + codeLine)

        const locations = byLines.get(lineIndex + 1) ?? []
        let locLines = standardFormatLocations(locations, offset)

        if (locLines.length >= 2) {
            // attempt a more compact format
            const compact = compactFormatLocations(locations, offset)
            if (compact) {
                locLines = compact
            }
        }

        outputLines.push(...locLines.map((l) => l.trimEnd()))
    })

    return "Code:\n" + outputLines.join("\n") + "\n\n"
}

function formatOutput(output: string | undefined, fixable: boolean): string {
    if (output === undefined) {
        return fixable ? "Output: unchanged\n\n" : ""
    }

    const linesFormatted = splitLines(output).map((l) => "  " + l)
    return `Output:\n${linesFormatted.join("\n")}\n\n`
}

function formatMessages({ code }: TestCase, messages: readonly Linter.LintMessage[]): string {
    if (messages.length === 0) {
        return "No errors\n"
    }

    return (
        messages
            .map(({ message, suggestions }) => {
                return {
                    message,
                    suggestions: suggestions?.map(({ desc, fix }) => {
                        const output =
                            code.slice(0, fix.range[0]) + fix.text + code.slice(fix.range[1])
                        return { desc, output: output + "\n" }
                    }),
                }
            })
            .map((obj, i) => {
                let content = obj.message
                if (obj.suggestions) {
                    content +=
                        "\n" +
                        YAML.stringify({ suggestions: obj.suggestions }, yamlOptions).replace(
                            /\n$/,
                            "",
                        )
                }
                const prefix = `[${i + 1}] `
                const space = " ".repeat(prefix.length)
                return (
                    prefix +
                    splitLines(content)
                        .map((l, j) => (j === 0 ? l : space + l))
                        .join("\n")
                )
            })
            .join("\n") + "\n"
    )
}
