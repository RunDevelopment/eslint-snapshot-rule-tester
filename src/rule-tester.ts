/**
 * This is a modified version of the ESLint's `RuleTester` class.
 *
 * @license MIT
 * Copyright OpenJS Foundation and other contributors, <www.openjsf.org>
 */

import { Linter, Rule, RuleTester } from "eslint"
import assert from "assert"
import { cloneDeeplyExcludesParent, freezeDeeply, getRuleOptionsSchema, sanitize } from "./util"
import { TestCase, TestCases, ruleTesterParameters, toTestCase } from "./test-case"
import deepEqual from "fast-deep-equal"
import { format } from "./format"
import { defaultConfig } from "./config"
import createAjv from "./ajv"
import { SnapshotTester } from "./snapshot/tester"
import { MochaSnapshotTester } from "./snapshot/mocha"
import stringify from "json-stable-stringify-without-jsonify"

const ajv = createAjv({ strictDefaults: true })

const testEnv = {
    get describe(): (name: string, fn: () => void) => void {
        return RuleTester["describe" as never]
    },
    get it(): (name: string, fn: () => void) => void {
        return RuleTester["it" as never]
    },
    get itOnly(): (name: string, fn: () => void) => void {
        return RuleTester["itOnly" as never]
    },
}

/**
 * Check if this test case is a duplicate of one we have seen before.
 * @param item test case object
 * @param seenTestCases set of serialized test cases we have seen so far (managed by this function)
 */
function checkDuplicateTestCase(item: TestCase, seenTestCases: Set<string>): void {
    let serialized
    try {
        // serialize and hope for the best
        serialized = stringify(item)
    } catch (error) {
        return
    }

    if (seenTestCases.has(serialized)) {
        assert.fail("detected duplicate test case")
    }
    seenTestCases.add(serialized)
}

interface ErrorReport {
    test: TestCase
    messages: Linter.LintMessage[]
    output: string | undefined
    filename: string | undefined
    configs: unknown
    beforeAST: unknown
    afterAST: unknown
}

const metaSchemaDescription = `
\t- If the rule has options, set \`meta.schema\` to an array or non-empty object to enable options validation.
\t- If the rule doesn't have options, omit \`meta.schema\` to enforce that no options can be passed to the rule.
\t- You can also set \`meta.schema\` to \`false\` to opt-out of options validation (not recommended).

\thttps://eslint.org/docs/latest/extend/custom-rules#options-schemas
`

export class SnapshotRuleTester {
    /**
     * The configuration to use for this tester. Combination of the tester
     * configuration and the default configuration.
     */
    private testerConfig: Linter.FlatConfig[]
    private linter: Linter

    constructor(testerConfig: Linter.FlatConfig = {}) {
        this.testerConfig = [testerConfig, { rules: { "rule-tester/validate-ast": "error" } }]
        this.linter = new Linter({ configType: "flat" })
    }

    public run(ruleName: string, rule: Rule.RuleModule, test: TestCases): void {
        const testerConfig = this.testerConfig
        const linter = this.linter
        const ruleId = `rule-to-test/${ruleName}`
        const validTestCases = new Set<string>()
        const invalidTestCases = new Set<string>()

        const baseConfig: Linter.FlatConfig[] = [
            { files: ["**"] }, // Make sure the default config matches for all files
            {
                plugins: {
                    "rule-to-test": {
                        rules: {
                            [ruleName]: {
                                ...rule,
                                // Create a wrapper rule that freezes the `context` properties.
                                create(context) {
                                    freezeDeeply(context.options)
                                    freezeDeeply(context.settings)
                                    freezeDeeply(context.parserOptions)

                                    // freezeDeeply(context.languageOptions);

                                    return rule.create(context)
                                },
                            },
                        },
                    },
                },
                languageOptions: {
                    ...defaultConfig[0].languageOptions,
                },
            },
            ...defaultConfig.slice(1),
        ]

        /**
         * Run the rule for the given item
         * @param item Item to run the rule against
         * @throws {Error} If an invalid schema.
         * @returns Eslint run result
         * @private
         */
        function runRuleForItem(item: TestCase): ErrorReport {
            if (item.filename) {
                // TODO: figure out what basePath does
                // flatConfigArrayOptions.basePath = path.parse(item.filename).root
            }

            const configs: Linter.FlatConfig[] = [...baseConfig, ...testerConfig]

            const code = item.code
            const filename = item.filename

            /*
             * Assumes everything on the item is a config except for the
             * parameters used by this tester
             */
            const itemConfig = { ...item }

            for (const parameter of ruleTesterParameters) {
                delete itemConfig[parameter]
            }

            // wrap any parsers
            const parser = itemConfig.languageOptions?.parser
            if (parser && typeof parser !== "object") {
                throw new Error(
                    "Parser must be an object with a parse() or parseForESLint() method.",
                )
            }

            /*
             * Create the config object from the tester config and this item
             * specific configurations.
             */
            configs.push(itemConfig)

            let ruleConfig: Linter.RuleEntry = 1
            if (item.options !== undefined) {
                ruleConfig = [1, ...item.options]
            }

            configs.push({
                rules: {
                    [ruleId]: ruleConfig,
                },
            })

            let schema

            try {
                schema = getRuleOptionsSchema(rule)
            } catch (err) {
                if (err instanceof Error) {
                    err.message += metaSchemaDescription
                }
                throw err
            }

            /*
             * Check and throw an error if the schema is an empty object (`schema:{}`), because such schema
             * doesn't validate or enforce anything and is therefore considered a possible error. If the intent
             * was to skip options validation, `schema:false` should be set instead (explicit opt-out).
             *
             * For this purpose, a schema object is considered empty if it doesn't have any own enumerable string-keyed
             * properties. While `ajv.compile()` does use enumerable properties from the prototype chain as well,
             * it caches compiled schemas by serializing only own enumerable properties, so it's generally not a good idea
             * to use inherited properties in schemas because schemas that differ only in inherited properties would end up
             * having the same cache entry that would be correct for only one of them.
             *
             * At this point, `schema` can only be an object or `null`.
             */
            if (schema && Object.keys(schema).length === 0) {
                throw new Error(`\`schema: {}\` is a no-op${metaSchemaDescription}`)
            }

            /*
             * Setup AST getters.
             * The goal is to check whether or not AST was modified when
             * running the rule under test.
             */
            let beforeAST, afterAST
            configs.push({
                plugins: {
                    "rule-tester": {
                        rules: {
                            "validate-ast": {
                                create() {
                                    return {
                                        Program(node) {
                                            beforeAST = cloneDeeplyExcludesParent(node)
                                        },
                                        "Program:exit"(node) {
                                            afterAST = node
                                        },
                                    }
                                },
                            },
                        },
                    },
                },
            })

            if (schema) {
                ajv.validateSchema(schema)

                if (ajv.errors) {
                    const errors = ajv.errors
                        .map((error) => {
                            const field =
                                error.dataPath[0] === "." ? error.dataPath.slice(1) : error.dataPath

                            return `\t${field}: ${error.message}`
                        })
                        .join("\n")

                    throw new Error([`Schema for rule ${ruleName} is invalid:`, errors].join("\n"))
                }

                /*
                 * `ajv.validateSchema` checks for errors in the structure of the schema (by comparing the schema against a "meta-schema"),
                 * and it reports those errors individually. However, there are other types of schema errors that only occur when compiling
                 * the schema (e.g. using invalid defaults in a schema), and only one of these errors can be reported at a time. As a result,
                 * the schema is compiled here separately from checking for `validateSchema` errors.
                 */
                try {
                    ajv.compile(schema)
                } catch (err) {
                    throw new Error(
                        `Schema for rule ${ruleName} is invalid: ${err instanceof Error ? err.message : String(err)}`,
                    )
                }
            }

            // Verify the code.
            const messages = linter.verify(code, configs, filename)
            const fatalErrorMessage = messages.find((m) => m.fatal)
            assert(
                !fatalErrorMessage,
                `A fatal parsing error occurred: ${fatalErrorMessage && fatalErrorMessage.message}`,
            )

            // Get output
            let output = undefined
            if (messages.some((m) => m.fix)) {
                const fixReport = linter.verifyAndFix(code, configs, filename)

                if (fixReport.fixed && fixReport.output !== code) {
                    output = fixReport.output

                    const errorMessageInFix = linter
                        .verify(output, configs, filename)
                        .find((m) => m.fatal)

                    assert(
                        !errorMessageInFix,
                        [
                            "A fatal parsing error occurred in autofix.",
                            `Error: ${errorMessageInFix && errorMessageInFix.message}`,
                            "Autofix output:",
                            output,
                        ].join("\n"),
                    )
                }
            }

            return {
                test: item,
                messages,
                output,
                beforeAST,
                afterAST: cloneDeeplyExcludesParent(afterAST),
                configs,
                filename,
            }
        }

        function formatReport(report: ErrorReport): string {
            return format(report.test, !!rule.meta?.fixable, report.output, report.messages)
        }

        /**
         * Check if the AST was changed
         * @param beforeAST AST node before running
         * @param afterAST AST node after running
         * @private
         */
        function assertASTDidNotChange(report: ErrorReport): void {
            if (!deepEqual(report.beforeAST, report.afterAST)) {
                assert.fail("Rule should not modify AST.")
            }
        }

        /**
         * Check if the template is valid or not
         * all valid cases go through this
         * @param item Item to run the rule against
         * @private
         */
        function testValidTemplate(item: TestCase): void {
            checkDuplicateTestCase(item, validTestCases)

            const result = runRuleForItem(item)
            const messages = result.messages

            if (messages.length > 0) {
                assert.strictEqual(
                    messages.length,
                    0,
                    `Should have no errors but had ${messages.length}. Details:\n\n` +
                        formatReport(result),
                )
            }

            assertASTDidNotChange(result)
        }

        /**
         * Check if the template is invalid or not
         * all invalid cases go through this.
         * @param item Item to run the rule against
         * @private
         */
        function testInvalidTemplate(item: TestCase, tester: SnapshotTester): void {
            checkDuplicateTestCase(item, invalidTestCases)

            const result = runRuleForItem(item)
            const messages = result.messages

            for (const message of messages) {
                if (message.suggestions) {
                    const seenMessageIndices = new Map<string, number>()

                    for (let i = 0; i < message.suggestions.length; i += 1) {
                        const suggestionMessage = message.suggestions[i].desc
                        const previous = seenMessageIndices.get(suggestionMessage)

                        assert.ok(
                            !seenMessageIndices.has(suggestionMessage),
                            `Suggestion message '${suggestionMessage}' reported from suggestion ${i} was previously reported by suggestion ${previous}. Suggestion messages should be unique within an error.`,
                        )
                        seenMessageIndices.set(suggestionMessage, i)
                    }
                }
            }

            const snapMessage = formatReport(result)
            tester.assertEqual(snapMessage)

            assertASTDidNotChange(result)
        }

        /*
         * This creates a mocha test suite and pipes all supplied info through
         * one of the templates above.
         * The test suites for valid/invalid are created conditionally as
         * test runners (eg. vitest) fail for empty test suites.
         */
        testEnv.describe(ruleName, () => {
            if (test.valid.length > 0) {
                testEnv.describe("valid", () => {
                    test.valid.map(toTestCase).forEach((valid) => {
                        testEnv[valid.only ? "itOnly" : "it"](
                            sanitize(valid.name || valid.code),
                            function () {
                                testValidTemplate(valid)
                            },
                        )
                    })
                })
            }

            if (test.invalid.length > 0) {
                testEnv.describe("invalid", () => {
                    const knownTitles = new Set<string>()
                    test.invalid.map(toTestCase).forEach((invalid) => {
                        let title = sanitize(invalid.name || invalid.code)

                        // ensure a unique title
                        if (knownTitles.has(title)) {
                            for (let i = 2; ; ++i) {
                                const newTitle = `${title} ${i}`
                                if (!knownTitles.has(newTitle)) {
                                    title = newTitle
                                    break
                                }
                            }
                        }
                        knownTitles.add(title)

                        testEnv[invalid.only ? "itOnly" : "it"](title, function (this: unknown) {
                            testInvalidTemplate(invalid, MochaSnapshotTester.parseContext(this)!)
                        })
                    })
                })
            }
        })
    }
}
