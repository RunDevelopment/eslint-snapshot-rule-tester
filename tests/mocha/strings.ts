import { Rule } from "eslint"
import { SnapshotRuleTester } from "../../src/rule-tester"

const rule: Rule.RuleModule = {
    meta: {
        fixable: "code",
        hasSuggestions: true,
        type: "suggestion",
        messages: {
            string: "Found a string. Use `{{expr}}` instead.",
            template: "Found a template string. Use `{{expr}}` instead.",
            somethingElse: "Use something else instead.",
        },
    },
    schema: [],
    create(context) {
        return {
            Literal(node) {
                if (typeof node.value === "string") {
                    context.report({
                        node,
                        messageId: "string",
                        data: { expr: "foo" },
                        fix(fixer) {
                            return fixer.replaceText(node, "foo")
                        },
                    })
                }
            },
            TemplateLiteral(node) {
                context.report({
                    node,
                    messageId: "template",
                    data: { expr: "bar" },
                    suggest: [
                        {
                            messageId: "somethingElse",
                            fix(fixer) {
                                return fixer.replaceText(node, "somethingElse")
                            },
                        },
                    ],
                })
            },
        }
    },
}

const tester = new SnapshotRuleTester({
    languageOptions: {
        ecmaVersion: "latest",
        sourceType: "module",
    },
})

tester.run("strings", rule, {
    valid: [
        String.raw`1 + 2`,
        String.raw`
class Foo {
    constructor() {
        this.bar = 12
    }
}
`.trim(),
    ],
    invalid: [
        String.raw`1 + "foo"`,
        String.raw`
class Foo {
    foo = "foo"
    constructor() {
        this.bar = "bar"
    }
}
`.trim(),
        String.raw`["foo", "bar"].join("\n")`,
        '`foo ${a ? "bar" : "baz"} ${`_${""}_`}`',
        '`\nfoo ${a ? "bar" : "baz"}\n${`_${""}_`}\n` + "1"+"2"',
        "`foo${1 + 2}`",
    ],
})
