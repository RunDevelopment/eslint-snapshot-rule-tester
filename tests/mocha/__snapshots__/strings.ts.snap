# eslint-snapshot-rule-tester format: v1


strings >> invalid >> 1 + "foo"
Code:
  1 | 1 + "foo"
    |     ^~~~~ [1]

Output:
  1 | 1 + foo

[1] Found a string. Use `foo` instead.
---


"strings >> invalid >> class Foo {\n    foo = \"foo\"\n    constructor() {\n        this.bar = \"bar\"\n    }\n}"
Code:
  1 | class Foo {
  2 |     foo = "foo"
    |           ^~~~~ [1]
  3 |     constructor() {
  4 |         this.bar = "bar"
    |                    ^~~~~ [2]
  5 |     }
  6 | }

Output:
  1 | class Foo {
  2 |     foo = foo
  3 |     constructor() {
  4 |         this.bar = foo
  5 |     }
  6 | }

[1] Found a string. Use `foo` instead.
[2] Found a string. Use `foo` instead.
---


strings >> invalid >> ["foo", "bar"].join("\n")
Code:
  1 | ["foo", "bar"].join("\n")
    |  ^~~~~  ^~~~~       ^~~~
    |  [1]    [2]         [3]

Output:
  1 | [foo, foo].join(foo)

[1] Found a string. Use `foo` instead.
[2] Found a string. Use `foo` instead.
[3] Found a string. Use `foo` instead.
---


strings >> invalid >> `foo ${a ? "bar" : "baz"} ${`_${""}_`}`
Code:
  1 | `foo ${a ? "bar" : "baz"} ${`_${""}_`}`
    | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [1]
    |            ^~~~~ [2]        ^~~~~~~~~ [4]
    |                    ^~~~~ [3]    ^~ [5]

Output:
  1 | `foo ${a ? foo : foo} ${`_${foo}_`}`

[1] Found a template string. Use `bar` instead.
    Suggestions:
      - Use something else instead.
        Output:
          1 | somethingElse
[2] Found a string. Use `foo` instead.
[3] Found a string. Use `foo` instead.
[4] Found a template string. Use `bar` instead.
    Suggestions:
      - Use something else instead.
        Output:
          1 | `foo ${a ? "bar" : "baz"} ${somethingElse}`
[5] Found a string. Use `foo` instead.
---


"strings >> invalid >> `\nfoo ${a ? \"bar\" : \"baz\"}\n${`_${\"\"}_`}\n` + \"1\"+\"2\""
Code:
  1 | `
    | ^\ [1]
  2 | foo ${a ? "bar" : "baz"}
    | ~~~~~~~~~~~~~~~~~~~~~~~~\ [1]
    |           ^~~~~ [2]
    |                   ^~~~~ [3]
  3 | ${`_${""}_`}
    | ~~~~~~~~~~~~\ [1]
    |   ^~~~~~~~~ [4]
    |       ^~ [5]
  4 | ` + "1"+"2"
    | ~ [1]   ^~~ [7]
    |     ^~~ [6]

Output:
  1 | `
  2 | foo ${a ? foo : foo}
  3 | ${`_${foo}_`}
  4 | ` + foo+foo

[1] Found a template string. Use `bar` instead.
    Suggestions:
      - Use something else instead.
        Output:
          1 | somethingElse + "1"+"2"
[2] Found a string. Use `foo` instead.
[3] Found a string. Use `foo` instead.
[4] Found a template string. Use `bar` instead.
    Suggestions:
      - Use something else instead.
        Output:
          1 | `
          2 | foo ${a ? "bar" : "baz"}
          3 | ${somethingElse}
          4 | ` + "1"+"2"
[5] Found a string. Use `foo` instead.
[6] Found a string. Use `foo` instead.
[7] Found a string. Use `foo` instead.
---


strings >> invalid >> `foo${1 + 2}`
Code:
  1 | `foo${1 + 2}`
    | ^~~~~~~~~~~~~ [1]

Output: unchanged

[1] Found a template string. Use `bar` instead.
    Suggestions:
      - Use something else instead.
        Output:
          1 | somethingElse
---
